Objective-C/Tests/ReplicatorTest+CustomConflict.m

```diff
@@ -233,6 +233,129 @@ - (void) testConflictResolverMergeDoc {
+                [errors addObject: err];
+        }];
+    }];
+    AssertEqual(errors.lastObject.code, CBLErrorConflict);
+    AssertEqualObjects(errors.lastObject.domain, CBLErrorDomain);
+    AssertEqualObjects([self.db documentWithID: docId].toDictionary, localData);
+    [replicator removeChangeListenerWithToken: token];
+    
+    // should be solved when the replicator runs next time!!
+    resolver = [[TestConflictResolver alloc] initWithResolver: ^CBLDocument* (CBLConflict* con) {
+        return con.remoteDocument;
+    }];
+    pullConfig.conflictResolver = resolver;
+    [self run: pullConfig errorCode: 0 errorDomain: nil];
+    AssertEqualObjects([self.db documentWithID: docId].toDictionary, remoteData);
+}
+
+// TODO: disable test since the unresolved conflict side is still under development.
+- (void) _testConflictResolverDifferentDBDoc {
+    CBLDatabase.log.console.domains = kCBLLogDomainAll;
+    CBLDatabase.log.console.level = kCBLLogLevelInfo;
+    NSString* docId = @"doc";
+    NSDictionary* localData = @{@"key1": @"value1"};
+    NSDictionary* remoteData = @{@"key2": @"value2"};
+    [self makeConflictFor: docId withLocal: localData withRemote: remoteData];
+    TestConflictResolver* resolver;
+    CBLReplicatorConfiguration* pullConfig = [self pullConfig];
+    
+    __weak CBLDatabase* weakOtherDB = otherDB;
+    resolver = [[TestConflictResolver alloc] initWithResolver: ^CBLDocument* (CBLConflict* con) {
+        return [weakOtherDB documentWithID: con.localDocument.id]; // doc from different DB!!
+    }];
+    pullConfig.conflictResolver = resolver;
+    
+    // make sure resolver is thrown the exception and skips the resolution.
+    __block id<CBLListenerToken> token;
+    __block CBLReplicator* replicator;
+    __block NSMutableArray<NSError*>* errors = [NSMutableArray array];
+    [self run: pullConfig reset: NO errorCode: 0 errorDomain: nil onReplicatorReady: ^(CBLReplicator* r) {
+        token = [r addDocumentReplicationListener: ^(CBLDocumentReplication* docRepl) {
+            NSError* err = docRepl.documents.firstObject.error;
+            if (err)
+                [errors addObject: err];
+        }];
+    }];
+    AssertEqual(errors.lastObject.code, CBLErrorConflict);
+    AssertEqualObjects(errors.lastObject.domain, CBLErrorDomain);
+    AssertEqualObjects([self.db documentWithID: docId].toDictionary, localData);
+    [replicator removeChangeListenerWithToken: token];
+    
+    // should be solved when the replicator runs next time!!
+    resolver = [[TestConflictResolver alloc] initWithResolver: ^CBLDocument* (CBLConflict* con) {
+        return con.remoteDocument;
+    }];
+    pullConfig.conflictResolver = resolver;
+    [self run: pullConfig errorCode: 0 errorDomain: nil];
+    AssertEqualObjects([self.db documentWithID: docId].toDictionary, remoteData);
+}
+
+// TODO: disable test since the unresolved conflict side is still under development.
+- (void) _testConflictResolverThrowingException {
```

Comment:
this test is causing memory leak. which says the [localDoc](https://github.com/couchbase/couchbase-lite-ios/blob/master/Objective-C/CBLDatabase.mm#L1028) and [remoteDoc](https://github.com/couchbase/couchbase-lite-ios/blob/master/Objective-C/CBLDatabase.mm#L1034) which are created is not getting removed form memory. Tried assigning them in the [catch](https://github.com/couchbase/couchbase-lite-ios/blob/master/Objective-C/CBLDatabase.mm#L1068), as well as finally, but no luck. Is this due to objc program interruption and causing ARC not to release some scope variables?

Reply:
No, ARC should work correctly with exceptions.
You can find what's still referencing those objects by turning on Malloc Stack logging (in the scheme 'Run' panel, Diagnostics tab, you have to turn off sanitizers first). Set a breakpoint after the time the objects are leaked, then in the Debug navigator set the rightmost popup to "View Memory Graph Hierarchy". Then you can find instances of CBLDocument and it'll show you a graph of which other objects reference them.

Reply:
used the technique, and could see that the CBLConflict and its properties(localDocument and remoteDocument) are causing it. CBLConflict was holding a strong reference to them, which was not released in case of exception(which i am still not clear why only for exception, since it is not happening in other unit tests).

I made them `__weak`, so now memory issue is gone.

Reply:
making `__weak` seems to be wrong. so updated with strong. and disabled test since the objc seems to be not exception-safe. which is ignoring the retained variable when exception happens. Added a TODO in the so that, if we can handle an expected memory warning for any future test. 

[clang.llvm.org - doc](http://clang.llvm.org/docs/AutomaticReferenceCounting.html#exceptions)